include/irr/asset/material_compiler/IR.h

```diff
@@ -0,0 +1,429 @@
+#include <irr/asset/ICPUSampler.h>
+#include <irr/core/alloc/LinearAddressAllocator.h>
+
+namespace irr {
+namespace asset {
+namespace material_compiler
+{
+
+class IR : public core::IReferenceCounted
+{
+    class SBackingMemManager
+    {
+        _IRR_STATIC_INLINE_CONSTEXPR size_t INITIAL_MEM_SIZE = 1ull<<20;
+        _IRR_STATIC_INLINE_CONSTEXPR size_t MAX_MEM_SIZE = 1ull<<20;
+        _IRR_STATIC_INLINE_CONSTEXPR size_t ALIGNMENT = _IRR_SIMD_ALIGNMENT;
+
+        uint8_t* mem;
+        size_t currSz;
+        core::LinearAddressAllocator<uint32_t> addrAlctr;
+
+    public:
+        SBackingMemManager() : currSz(INITIAL_MEM_SIZE), addrAlctr(nullptr, 0u, 0u, ALIGNMENT, MAX_MEM_SIZE) {
+            mem = reinterpret_cast<uint8_t*>(_IRR_ALIGNED_MALLOC(currSz, ALIGNMENT));
+        }
+        ~SBackingMemManager() {
+            _IRR_ALIGNED_FREE(mem);
+        }
+
+        uint8_t* alloc(size_t bytes)
+        {
+            auto addr = addrAlctr.alloc_addr(bytes, ALIGNMENT);
+            //TODO reallocation will invalidate all pointers to nodes, so...
+            //1) never reallocate (just have reasonably big buffer for nodes)
+            //2) make some node_handle class that will work as pointer but is based on offset instead of actual address
+            if (addr+bytes > currSz) {
+                size_t newSz = currSz<<1;
+                if (newSz > MAX_MEM_SIZE) {
+                    addrAlctr.free_addr(addr, bytes);
+                    return nullptr;
+                }
+
+                void* newMem = _IRR_ALIGNED_MALLOC(newSz, ALIGNMENT);
+                memcpy(newMem, mem, currSz);
+                _IRR_ALIGNED_FREE(mem);
+                mem = reinterpret_cast<uint8_t*>(newMem);
+                currSz = newSz;
+            }
+
+            return mem+addr;
+        }
+    };
+
+protected:
+    ~IR()
+    {
+        //call destructors on all nodes
+        for (auto* root : roots)
+        {
+            core::stack<decltype(root)> s;
+            s.push(root);
+            while (!s.empty())
+            {
+                auto* n = s.top();
+                s.pop();
+                for (auto* c : n->children)
+                    s.push(c);
+
+                if (!n->deinited) {
+                    n->~INode();
+                    n->deinited = true;
+                }
+            }
+        }
+    }
+
+public:
+    IR() : memMgr() {}
+
+    void deinitTmpNodes()
+    {
+        for (INode* n : tmp)
+            n->~INode();
+        tmp.clear();
+    }
+
+    template <typename NodeType, typename ...Args>
+    NodeType* allocNode(Args&& ...args)
+    {
+        uint8_t* ptr = memMgr.alloc(sizeof(NodeType));
+        return new (ptr) NodeType(std::forward<Args>(args)...);
+    }
+    template <typename NodeType, typename ...Args>
+    NodeType* allocRootNode(Args&& ...args)
+    {
+        auto* root = allocNode<NodeType>(std::forward<Args>(args)...);
+        roots.push_back(root);
+        return root;
+    }
```

Comment:
make sure to leave a comment that we can have multiple root nodes ?
(or do you have multiple IR for material?)